# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/angle/checkout/src/libANGLE/Observer.h
# Commit: 13282d7a47a5
# Full Hash: 13282d7a47a57a73be8d060efd9645b4d642f03e
# Author: Miko Mynttinen <mikokm@gmail.com>
# Date: 2019-09-08 09:46:52
# Regressor Bug: 1578576
# File Overlap Count: 4
# Description:
#   Bug 1578576 - Part 2: Update to ANGLE 3865 r=jgilbert
#   
#   Depends on D44561
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D44579
# ==============================================================================

diff -r 3064469c073d -r 13282d7a47a5 gfx/angle/checkout/src/libANGLE/Observer.h
--- a/gfx/angle/checkout/src/libANGLE/Observer.h	Sat Sep 07 13:36:46 2019 +0000
+++ b/gfx/angle/checkout/src/libANGLE/Observer.h	Sat Sep 07 13:38:36 2019 +0000
@@ -16,11 +16,6 @@
 #include "common/FastVector.h"
 #include "common/angleutils.h"
 
-namespace gl
-{
-class Context;
-}  // namespace gl
-
 namespace angle
 {
 template <typename HaystackT, typename NeedleT>
@@ -31,14 +26,31 @@
 
 using SubjectIndex = size_t;
 
+// Messages are used to distinguish different Subject events that get sent to a single Observer.
+// It could be possible to improve the handling by using different callback functions instead
+// of a single handler function. But in some cases we want to share a single binding between
+// Observer and Subject and handle different types of events.
 enum class SubjectMessage
 {
-    CONTENTS_CHANGED,
-    STORAGE_CHANGED,
-    BINDING_CHANGED,
-    DEPENDENT_DIRTY_BITS,
-    RESOURCE_MAPPED,
-    RESOURCE_UNMAPPED,
+    // Used by gl::VertexArray to notify gl::Context of a gl::Buffer binding count change. Triggers
+    // a validation cache update.
+    BindingChanged,
+
+    // Only the contents (pixels, bytes, etc) changed in this Subject. Distinct from the object
+    // storage.
+    ContentsChanged,
+
+    // Sent by gl::Sampler, gl::Texture, gl::Framebuffer and others to notifiy gl::Context. This
+    // flag indicates to call syncState before next use.
+    DirtyBitsFlagged,
+
+    // Generic state change message. Used in multiple places for different purposes.
+    SubjectChanged,
+
+    // Indicates a bound gl::Buffer is now mapped or unmapped. Passed from gl::Buffer, through
+    // gl::VertexArray, into gl::Context. Used to track validation.
+    SubjectMapped,
+    SubjectUnmapped,
 };
 
 // The observing class inherits from this interface class.
@@ -46,9 +58,7 @@
 {
   public:
     virtual ~ObserverInterface();
-    virtual void onSubjectStateChange(const gl::Context *context,
-                                      SubjectIndex index,
-                                      SubjectMessage message) = 0;
+    virtual void onSubjectStateChange(SubjectIndex index, SubjectMessage message) = 0;
 };
 
 class ObserverBindingBase
@@ -76,7 +86,7 @@
     Subject();
     virtual ~Subject();
 
-    void onStateChange(const gl::Context *context, SubjectMessage message) const;
+    void onStateChange(SubjectMessage message) const;
     bool hasObservers() const;
     void resetObservers();
 
@@ -112,7 +122,7 @@
 
     ANGLE_INLINE void reset() { bind(nullptr); }
 
-    void onStateChange(const gl::Context *context, SubjectMessage message) const;
+    void onStateChange(SubjectMessage message) const;
     void onSubjectReset() override;
 
     ANGLE_INLINE const Subject *getSubject() const { return mSubject; }